### PR TITLE
Add single-node Qwen3 DAPO math example

### PR DESCRIPTION
**Pull request by felipemello1**
## Summary

- Add a single-node Qwen3-4B-Base DAPO-Math example with a TP=2 trainer and six independent TP=1 generators.
- Train on the 12,643-row filtered DAPO-Math dataset and validate on AIME 2025 with Math-Verify rewards.
- Document the recipe and the measured 150-step training result.
- Smoke run: https://meta.wandb.io/felipemello/titan_rl/runs/3xhecewt

image
image

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -18,6 +18,9 @@ dependencies = [
     "tokenizers>=0.15.0",
     "safetensors",
 
+    # Rule-based verification for math RL examples
+    "math-verify==0.9.0",
```
Comment (tianyu-l) [inline pyproject.toml]: This should be recipe / example level of requirements, rather than repo-wise? Some one who works on pre-training / diffusion models may not need this lib.

If you need to put this in CI, we can do something similar to https://github.com/pytorch/torchtitan/blob/main/.ci/docker/requirements-vlm.txt

File: torchtitan/experiments/rl/examples/dapo_math/run_train.sh
Comment (tianyu-l) [inline torchtitan/experiments/rl/examples/dapo_math/run_train.sh:1]: Do we need this? How should I appreciate this vs. https://github.com/felipemello1/torchtitan/blob/main/torchtitan/experiments/rl/train.py

File: torchtitan/experiments/rl/examples/dapo_math/README.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+## 150-step result
+
+The single-node TP=2 trial completed optimizer steps 0 through 149 in 1 hour 46 minutes. This trial used the earlier 7,168-token response and 9,216-token packing limits; the runnable configuration above restores the full 8,192-token response budget.
+
+```text
+metric                              first 10 steps    last 10 steps
+rollout reward, mean                0.067             0.401
+rollout total length, mean          988 tokens        3,284 tokens
+rollout truncation rate, mean       1.75%             14.55%
+```
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/examples/dapo_math/README.md]: Could you share a wandb log on the real job?

File: torchtitan/experiments/rl/examples/dapo_math/README.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+Use `rl_dapo_qwen3_4b_math_32k` as the config name to run the 32K variant.
+
+## 150-step reference result
+
+TODO: add eval results. Add 32k variant.
+
+![Qwen3-4B DAPO Math-Verify reward](./assets/qwen3_4b_7k_reward.png)
+
+![Qwen3-4B DAPO mean response length](./assets/qwen3_4b_7k_response_length.png)
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/examples/dapo_math/README.md:102]: Curves in these pictures would be easy to get outdated and not reproducible on later commit. Maybe only add them in PR summary, or add a commit pin to the results.

**Review by tianyu-l — APPROVED** [review]

File: torchtitan/experiments/rl/examples/dapo_math/README.md [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+Answer: \boxed{0}
+
+Reward: 1
+```
+
+## Datasets
+
+Training uses the 12,643-row [filtered DAPO-Math dataset](https://huggingface.co/datasets/hamishivi/DAPO-Math-17k-Processed_filtered). Each row contains one user prompt and its verifiable final answer.
+
+Validation uses all 30 problems from [AIME 2025](https://huggingface.co/datasets/opencompass/AIME2025). The same single-turn environment and Math-Verify reward are used for training and validation.
```
Comment (yichuan-w) [inline torchtitan/experiments/rl/examples/dapo_math/README.md:42]: can we also put a validation set accuracy figure in the readme so people can be more convinced

**Comment by yichuan-w** [comment]
LGTM, very clean PR

File: torchtitan/experiments/rl/examples/dapo_math/README.md (hunk 4 above)
Comment (felipemello1) [reply torchtitan/experiments/rl/examples/dapo_math/README.md:42]: yes, i ahve to rerun it. I left as a todo at the bottom of the readme
